### PR TITLE
support default printing in format intrinsic 

### PR DESCRIPTION
**Pull request by assem2002**
towards https://github.com/lfortran/lfortran/issues/2543.

We now support printing default formatting using `lcompilers_string_format_fortran()` so we can move forward towards closing the issue mentioned above, and also printing arrays in the backend. 

After finishing this one, we should be able to change `print_t` node to the suggested node mentioned in the issue https://github.com/lfortran/lfortran/issues/2543.

### CODE REVIEW

**Comment by assem2002** [comment]
Expected to fail due to reference files.
It would be a great help for anyone to check integration test `file_09.f90` it doesn't fail in main if you run with python script, but fails when you run it as stand alone file.

**Comment by kmr-srbh** [comment]
The issue with executing `file_09.f90` occurs on latest main also. In my experience, I have noticed this many times that some tests fail when executing by hand, but pass when executing through the python script.

**Comment by kmr-srbh** [comment]
I think we have a bug here related to reading into an `allocatable` array.

For the test `file_09.f90`, reading into a general array works correctly, but reading into an `allocatable` array leads to input of random garbage values.

```fortran
program file_09
    implicit none

    integer :: u = 11, i, j
    character :: arr(2, 2)
    character, allocatable :: alloc_arr(:, :)

    open(u, file="file_01_data.txt", form="unformatted", access="stream", status="old")
    read(u) arr
    close(u)

    print *, "arr(1, 1): ", arr(1, 1) ! added by me
    print *, "arr(1, 2): ", arr(1, 2) ! added by me
    if (arr(1, 1) /= '1' .or. arr(1, 2) /= '1') error stop
    print *, arr

    open(u, file="file_01_data.txt", form="unformatted", access="stream", status="old")
    allocate(alloc_arr(2, 2))
    read(u) alloc_arr
    close(u)

    print *, "alloc_arr(2, 1): ", alloc_arr(2, 1) ! added by me
    print *, "alloc_arr(2, 2): ", alloc_arr(2, 2) ! added by me
    if (alloc_arr(2, 1) /= '0' .or. alloc_arr(2, 2) /= '3') error stop
    print *, alloc_arr
end program file_09
```

### GFortran

```console
(lf) saurabh-kumar@Awadh:~/Projects/System/lfortran/integration_tests$ gfortran file_09.f90 && ./a.out
 arr(1, 1): 1
 arr(1, 2): 1
 1013
 alloc_arr(2, 1): 0
 alloc_arr(2, 2): 3
 1013
```

### LFortran

```console
(lf) saurabh-kumar@Awadh:~/Projects/System/lfortran/integration_tests$ lfortran file_09.f90 --no-warnings

arr(1, 1): 1
arr(1, 2): 1
11
03

alloc_arr(2, 1): 0@�5M}
alloc_arr(2, 2): 3;�5M}
ERROR STOP
```

> file_01_data.txt

```
10130
c
4.20
9223372036854775804
```

I am opening an issue for this with an MRE.

**Comment by certik** [comment]
> , I have noticed this many times that some tests fail when executing by hand, but pass when executing through the python script.

Ok, so this is a separate issue, but we need to investigate this. Can you create a dedicated issue just to this problem, and how how to reproduce it on your computer? Post the failure that you see locally, and the same failure then disappears when ran via the Python runner.

**Comment by kmr-srbh** [comment]
The above issue with reading occurs only when we read into an `allocatable` array after reading from a general array.

For example, this works correctly:

```fortran
program file_09
    implicit none

    integer :: u = 11
    character, allocatable :: alloc_arr(:, :)

    open(u, file="file_01_data.txt", form="unformatted", access="stream", status="old")
    allocate(alloc_arr(2, 2))
    read(u) alloc_arr
    close(u)

    print *, "alloc_arr(2, 1): ", alloc_arr(2, 1)
    print *, "alloc_arr(2, 2): ", alloc_arr(2, 2)
    if (alloc_arr(2, 1) /= '0' .or. alloc_arr(2, 2) /= '3') error stop
    print *, alloc_arr
end program file_09
```

### GFortran

```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran/integration_tests$ gfortran file_09.f90 && ./a.out
 alloc_arr(2, 1): 0
 alloc_arr(2, 2): 3
 1013
```

### LFortran

```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran/integration_tests$ lfortran file_09.f90 --no-warnings
alloc_arr(2, 1): 0
alloc_arr(2, 2): 3
11
03
```

File: tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.stderr [hunk 1]
```diff
@@ -1 +1,9 @@
-semantic error: The `runtime stacktrace` is not enabled. To get the stack traces or debugging information, please re-build LFortran with `-DWITH_RUNTIME_STACKTRACE=yes`
+code generation error: asr_to_llvm: module failed verification. Error:
```
Comment (kmr-srbh) [inline tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.stderr:1]: This generated reference is not correct. It had contained the generated stacktrace due to the error we test. @assem2002 I request you to look into this.

```console
  File "tests/errors/runtime_stacktrace_01.f90", line 3
    print *, main()
  File "tests/errors/runtime_stacktrace_01.f90", line 10
    print *, g()
  File "tests/errors/runtime_stacktrace_01.f90", line 20
    call f()
  File "tests/errors/runtime_stacktrace_01.f90", line 14
    stop 1
STOP 1
```

Comment (kmr-srbh) [reply tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.stderr:1]: Thanks! I do not see the updated error reference now. :+1:

**Comment by assem2002** [comment]
The test with `llvm_wasm-emcc` weirdly append to character with `len=20` (specially) an additional character that is always equal to decimal of `19`, Does this seem familiar?

**Comment by assem2002** [comment]
It's ready for review.

for the issue I have we emcc, This test adds a weird symbol to the end of variable `str` (using 'wasm_emcc' as llvm backend)
```fortran 
program intrinsics_177
    character(len=20) :: str = 'gfortran   '
    str = adjustr(str)
    print "(a)", str
end program
```
It prints correct output when you just change the len of  `str` to any number other than **20**

**Comment by certik** [comment]
Ah, if the issue is already with main, then just open up an issue with instructions how to reproduce. I think we can ignore this issue in this PR, unless it is somehow a blocker.

**Comment by assem2002** [comment]
Okay, I'll open up an issue for that. I've tested the same issue with older version of lfortran before I started refactoring `stringFormat` and `print` and the issue still persists.

**Comment by certik** [comment]
Does it make sense to modify the `Print(expr* values, expr? separator, expr? end)` ASR node to just be `Print(expr text)`? And enforce in verify that the `text` argument must be a Character ttype.

File: integration_tests/intrinsics_134.f90 [hunk 2]
```diff
@@ -2,7 +2,7 @@ program intrinsics_134
 
     character(4), parameter :: y = adjustl("okay")
     character(len=5) :: x
-    character(len=20) :: str = '   gfortran'
+    character(len=21) :: str = '   gfortran'
```
Comment (certik) [inline integration_tests/intrinsics_134.f90:5]: Is this change needed due to https://github.com/lfortran/lfortran/pull/4615#issuecomment-2289620741 ?

Unless it is needed for this PR, I would not make these changes here, but rather in a fix for that issue.

**Comment by assem2002** [comment]
> Does it make sense to modify the `Print(expr* values, expr? separator, expr? end)` ASR node to just be `Print(expr text)`? And enforce in verify that the `text` argument must be a Character ttype.

Yes, It should be either a `stringformat` or a `stringConstant`

File: integration_tests/intrinsics_134.f90 (hunk 2 above)
Comment (assem2002) [reply integration_tests/intrinsics_134.f90:5]: It's a workaround to pass the errors in 'LLVM->WASM' CI test, as `len =20` weirdly causes the error mentioned  here https://github.com/lfortran/lfortran/pull/4615#issuecomment-2289620741

Comment (certik) [reply integration_tests/intrinsics_134.f90:5]: The same error happens in master, but in there it doesn't seem to cause the CI to fail. Why is that?

**Comment by certik** [comment]
It can also be a runtime string, so any expression that returns a string (such as FunctionCall).

**Comment by assem2002** [comment]
Yeah, that's true. Anyways this PR wraps everything in `stringFormat` node so it's always a `character` type

**Review by certik — APPROVED** [review]
I think this is fine, but I have one issue with it:

Right now it seems the StringFormat is always used even if we are just printing a string. Is it possible to later refactor this to only put StringFormat in if needed, but not when not needed?

File: tests/reference/run-format_12-43152e4.stdout [hunk 3]
```diff
@@ -1,7 +1,7 @@
-testing output using formatting
+testing output using formatting
```
Comment (certik) [inline tests/reference/run-format_12-43152e4.stdout:1]: This white space might be a problem.

File: tests/reference/run-format_11-72f8b07.stdout [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
+test writing a logical    
 
-formatted write
+formatted write    
       T
       F
-using print
-True
-False
+using print
```
Comment (certik) [inline tests/reference/run-format_11-72f8b07.stdout:6]: Here is the corresponding Fortran code:
```fortran
program format_11
implicit none
logical :: b
print *, 'test writing a logical'
print *
b = .false.
print *, 'formatted write'
write (*, 700) .true.
write (*, 700) b
print *, 'using print'
print *, .true.
print *, b
stop
700 format (l7)
end program
```
So it should not be printing empty space at the end.

**Comment by certik** [comment]
I found one issue with empty spaces at the end: https://github.com/lfortran/lfortran/pull/4615/files#r1719177245, so we need to fix that before merging.

Why not also in AST->ASR to check if the argument has a type string to just leave it? Only if it is any other type, or a specific format is requested, to use StringFormat. That will greatly reduce the number of test changes, it would actually fix the bug above (or at least prevent the bug to show up). 

Otherwise I think this PR does what we need. For things like `print *, i` where `i` is an integer, it prints it using StringFormat, which is the way to go.

**Comment by assem2002** [comment]
I'll work on the requested changes, they're all doable.
For this expression --> `print *, "hello","world",1,2` should I separate the strings from integers so that integers would get wrapped in `stringFormat` node while strings a won't? or just stick with the case of print arguments being only strings like this --> `print *, "hello","world"`.

File: tests/reference/run-format_11-72f8b07.stdout (hunk 4 above)
Comment (assem2002) [reply tests/reference/run-format_11-72f8b07.stdout]: Done. by https://github.com/lfortran/lfortran/pull/4615/commits/bee5a2b8f2936025a8a6f96a7fe563c08ac590a8, https://github.com/lfortran/lfortran/pull/4615/commits/35a5c8588ea248bc52a98fd42adcd97abd059664

File: tests/reference/run-format_12-43152e4.stdout (hunk 3 above)
Comment (assem2002) [reply tests/reference/run-format_12-43152e4.stdout]: Done. by https://github.com/lfortran/lfortran/commit/bee5a2b8f2936025a8a6f96a7fe563c08ac590a8, https://github.com/lfortran/lfortran/commit/35a5c8588ea248bc52a98fd42adcd97abd059664

**Comment by certik** [comment]
`print *, "hello","world",1,2` should be `Print(StringFormat((), ["hello","world",1,2])`. Later, we can add various ASR passes that can transform StringFormat to various other forms, in this particular case since we know the formatting and even all the expressions at compile time, the optimizing ASR pass should replace it with "hello world 1 2" (or whatever the default formatting of StringFormat is). But all of this is for later.

Right now our job is to represent all these operations with StringFormat and the simplest possible way.

File: tests/reference/run-format1-04216cf.stdout [hunk 5]
```diff
@@ -1,4 +1,4 @@
-okb
+ok    b
```
Comment (certik) [inline tests/reference/run-format1-04216cf.stdout:1]: There is some issue here.

Comment (assem2002) [reply tests/reference/run-format1-04216cf.stdout:1]: I think this is compiler dependent thing, It corresponds to this expression `print *, "ok", "b"`. We add 4 spaces after every argument, while gfortran doesn't add spaces between consecutive strings arguments . If wish to do the same then it's doable.

**Comment by assem2002** [comment]
I think it's ready to merge

**Comment by certik** [comment]
@assem2002 awesome, thank you for this excellent PR. If anything breaks, we'll fix it with subsequent PRs.
